Source/AuthenticationInterceptor.swift

```diff
@@ -0,0 +1,324 @@
+        guard let credential = credential else {
+            let error = AuthenticationError.missingCredential
+            completion(.doNotRetryWithError(error))
+            return
+        }
+
+        // Retry the request if inspector verifies it was authenticated with a previous credential
+        guard authenticator.isRequest(urlRequest, authenticatedWith: credential) else {
+            completion(.retry)
+            return
+        }
+
+        $mutableState.write { mutableState in
+            mutableState.requestsToRetry.append(completion)
+
+            guard !mutableState.isRefreshing else { return }
+
+            refresh(credential, insideLock: &mutableState)
+        }
+    }
+
+    // MARK: Refresh
+
+    private func refresh(_ credential: AuthCredential, insideLock mutableState: inout MutableState) {
+        guard !isRefreshExcessive(insideLock: &mutableState) else {
+            let error = AuthenticationError.excessiveRefresh
+            handleRefreshFailure(error, insideLock: &mutableState)
+            return
+        }
+
+        mutableState.refreshTimestamps.append(Date())
+        mutableState.isRefreshing = true
+
+        authenticator.refresh(credential) { result in
+            self.$mutableState.write { mutableState in
+                switch result {
+                case .success(let credential):
+                    self.handleRefreshSuccess(credential, insideLock: &mutableState)
+
+                case .failure(let error):
+                    self.handleRefreshFailure(error, insideLock: &mutableState)
+                }
+            }
+        }
+    }
+
+    private func isRefreshExcessive(insideLock mutableState: inout MutableState) -> Bool {
+        guard let refreshWindow = mutableState.refreshWindow else { return false }
+
+        let refreshWindowMin = Date(timeIntervalSinceNow: -refreshWindow.interval)
+
+        let refreshAttemptsWithinWindow = mutableState.refreshTimestamps.reduce(into: 0) { attempts, refreshTimestamp in
+            guard refreshWindowMin <= refreshTimestamp else { return }
+            attempts += 1
+        }
+
+        let isRefreshExcessive = refreshAttemptsWithinWindow >= refreshWindow.maximumAttempts
+
+        return isRefreshExcessive
+    }
+
+    private func handleRefreshSuccess(_ credential: AuthCredential, insideLock mutableState: inout MutableState) {
+        mutableState.credential = credential
+
+        let adaptOperations = mutableState.adaptOperations
+        let requestsToRetry = mutableState.requestsToRetry
+
+        mutableState.adaptOperations.removeAll()
+        mutableState.requestsToRetry.removeAll()
+
+        mutableState.isRefreshing = false
+
+        // Dispatch to queue to hop out of the mutable state lock
+        queue.async {
+            adaptOperations.forEach { self.adapt($0.urlRequest, for: $0.session, completion: $0.completion) }
+            requestsToRetry.forEach { $0(.retry) }
+        }
```

Comment:
There was no way around this @jshier because the `handleRefreshFailure` must be called inside the lock when excessive refresh is tripped. If we hop out of the lock and then call into `handleRefreshFailure`, things can get out-of-sync. The only real option in the end is to hop out of the mutable state lock when calling the queued completions. This only happens once when refresh is completed, so it shouldn't cause any performance issues since it's most likely on called every 60 minutes or so.